fabric_generator/utilities.py

```diff
@@ -0,0 +1,1079 @@
+                                belHasClockInput = True
+
+                    except:
+                        raise Exception(f"{wire[1]} file for BEL not found")
+
+                    if len(wire) > 2:
+                        prefix = wire[2]
+                    else:
+                        prefix = ""
+                    nports = []
+                    inputPorts = []
+                    outputPorts = []
+
+                    for port in ports[0]:
+                        nports.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                        # Also add to distinct input/output lists
+                        inputPorts.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                    for port in ports[1]:
+                        nports.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                        outputPorts.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                    cTile.belPorts.update(nports)
+
+                    belListWithIO.append(
+                        [wire[1][0:-5:], prefix, inputPorts, outputPorts, belHasClockInput])
+                    belList.append(
+                        [wire[1][0:-5:], prefix, nports, belHasClockInput])
+
+                elif wire[0] in ["NORTH", "SOUTH", "EAST", "WEST"]:
+                    # Wires are added in next pass - this pass generates port lists to be used for wire generation
+                    if wire[1] != "NULL":
+                        portList.append(wire[1])
+                    if wire[4] != "NULL":
+                        portList.append(wire[4])
+                    wireTextList.append({"direction": wire[0], "source": wire[1], "xoffset": wire[2],
+                                        "yoffset": wire[3], "destination": wire[4], "wire-count": wire[5]})
+                # We just treat JUMPs as normal wires - however they're only on one tile so we can add them directly
+                elif wire[0] == "JUMP":
+                    if "NULL" not in wire:
+                        wires.append({"direction": wire[0], "source": wire[1], "xoffset": wire[2],
+                                     "yoffset": wire[3], "destination": wire[4], "wire-count": wire[5]})
+            cTile.wires = wires
+            cTile.x = j
+            # cTile.y = archFabric.height - i -1
+            cTile.y = i
+
+            cTile.bels = belList
+            cTile.belsWithIO = belListWithIO
+            row.append(cTile)
+            portMap[cTile] = portList
+            wireMap[cTile] = wireTextList
+        archFabric.tiles.append(row)
+
+        # Add wires to model
+
+    for row in archFabric.tiles:
+        for tile in row:
+            wires = []
+            wireTextList = wireMap[tile]
+            tempAtomicWires = []
+            # Wires from tile
+            for wire in wireTextList:
```

Comment:
you are using `int(wire["xoffset"])` and `int(wire["yoffset"])` a lot, perhaps make some temporary variables for that rather than doing the array lookup and int conversion each time?